pkg/reconciler/broker/broker_test.go

```diff
@@ -0,0 +1,183 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	clientgotesting "k8s.io/client-go/testing"
+	"knative.dev/eventing/pkg/utils"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/client/injection/ducks/duck/v1/addressable"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	logtesting "knative.dev/pkg/logging/testing"
+	. "knative.dev/pkg/reconciler/testing"
+)
+
+const (
+	testNS     = "testnamespace"
+	brokerName = "test-broker"
+
+	testProject = "test-project-id"
+	generation  = 1
+	testUID     = "abc123"
+	systemNS    = "knative-testing"
+
+	finalizerName = "brokers.eventing.knative.dev"
+)
+
+var (
+	testKey = fmt.Sprintf("%s/%s", testNS, brokerName)
+
+	brokerFinalizerUpdatedEvent = Eventf(corev1.EventTypeNormal, "FinalizerUpdate", `Updated "test-broker" finalizers`)
+	brokerReconciledEvent       = Eventf(corev1.EventTypeNormal, "BrokerReconciled", `Broker reconciled: "testnamespace/test-broker"`)
+	brokerFinalizedEvent        = Eventf(corev1.EventTypeNormal, "BrokerFinalized", `Broker finalized: "testnamespace/test-broker"`)
+
+	brokerAddress = &apis.URL{
+		Scheme: "http",
+		Host:   fmt.Sprintf("%s.%s.svc.%s", ingressServiceName, systemNS, utils.GetClusterDomainName()),
+		Path:   fmt.Sprintf("/%s/%s", testNS, brokerName),
+	}
+)
+
+func init() {
+	// Add types to scheme
+	_ = brokerv1beta1.AddToScheme(scheme.Scheme)
+}
+
+func TestAllCases(t *testing.T) {
+	table := TableTest{{
+		Name: "bad workqueue key",
+		Key:  "too/many/parts",
+	}, {
+		Name: "key not found",
+		Key:  testKey,
+	}, {
+		Name: "Broker is being deleted, no topic or sub exists",
+		Key:  testKey,
+		Objects: []runtime.Object{
+			NewBroker(brokerName, testNS,
+				WithBrokerClass(brokerv1beta1.BrokerClass),
+				WithInitBrokerConditions,
+				WithBrokerDeletionTimestamp),
+		},
+		WantEvents: []string{
+			brokerFinalizedEvent,
+		},
+		OtherTestData: map[string]interface{}{
+			"ps": gpubsub.TestClientData{
+				TopicData: gpubsub.TestTopicData{
+					Exists: false,
+				},
+				SubscriptionData: gpubsub.TestSubscriptionData{
+					Exists: false,
+				},
+			},
+		},
+	}, {
+		Name: "Broker is being deleted, topic and sub exists",
+		Key:  testKey,
+		Objects: []runtime.Object{
+			NewBroker(brokerName, testNS,
+				WithBrokerClass(brokerv1beta1.BrokerClass),
+				WithBrokerUID(testUID),
+				WithInitBrokerConditions,
+				WithBrokerDeletionTimestamp),
+		},
+		WantEvents: []string{
+			Eventf(corev1.EventTypeNormal, "TopicDeleted", `Deleted PubSub topic "cre-bkr_testnamespace_test-broker_abc123"`),
+			Eventf(corev1.EventTypeNormal, "SubscriptionDeleted", `Deleted PubSub subscription "cre-bkr_testnamespace_test-broker_abc123"`),
+			brokerFinalizedEvent,
+		},
+		OtherTestData: map[string]interface{}{
+			"ps": gpubsub.TestClientData{
+				TopicData: gpubsub.TestTopicData{
+					Exists: true,
+				},
+				SubscriptionData: gpubsub.TestSubscriptionData{
+					Exists: true,
+				},
+			},
```

Comment:
Can the test case assert these were deleted in the test client?

Reply:
Looks like the test client only supports configuring responses, not asserting that methods were called. This would be a useful feature but I think it should be a separate issue.

Reply:
We should be able to use pstest instead of gclient and verify the state of resources in the test server.

Reply:
There is an issue for this https://github.com/google/knative-gcp/issues/781

Reply:
That issue covers existing tests and is a P2 meaning it may not get prioritized anytime soon. I'd prefer for new tests to test pubsub state when they are merged. I think pstest should be a fairly straightforward replacement for the test client. If it's not it would suffice instead to add some simple state to the test client to track topics and subscriptions.

Reply:
Ian is correct that #781 isn't clear that this is a testing gap now. I created #886 to track fixing this for the broker controller specifically. We can then copy the approach to other controllers.

Reply:
Thanks Grant for the follow up!